packages/2.questions-search/src/app/views/questions/question-list/question-list.component.ts

```diff
@@ -9,18 +13,38 @@ import { QuestionsService } from 'src/app/services/questions.service';
 })
 export class QuestionListComponent implements OnInit {
 
-  questions: Array<Question>;
-  constructor(questionsSvc: QuestionsService) {
+  //Used by the mat-table element
+  displayedColumns: string[] = ['id', 'name', 'email', 'delete'];
+  tableDataSource: MatTableDataSource<Question>;
+  @ViewChild(MatSort) sort: MatSort;
+
+  constructor(private questionsSvc: QuestionsService, private createQuestionDialog: MatDialog) {
     questionsSvc.questions$.subscribe(q => {
-      this.questions = q;
+      this.tableDataSource = new MatTableDataSource(q);
     });
```

Comment:
Nothing wrong with this!

Just a few things that would make this perfect. Beware of the hanging subscription! It leaks memory like a sieve. :P

Also, keep in mind that, although changing the reference does in fact re-render the component, you will lose the filters and sorting. You might want to create the `dataSource` once and then assign it the data. You also [need to re-render rows](https://material.angular.io/components/table/overview#1-write-your-mat-table-and-provide-data), in that case.

It would end up being something like this:
```typescript
questionsSvc.questions$
  .pipe(takeUntil(this.destroy$)) // this is to avoid hanging subscriptions. It kills them when `destroy$` fires.
  .subscribe(data => { this.tableDataSource.data = data; this.table.renderRows(); });
```

With `table` being:
```typescript
@ViewChild(MatTable) table: MatTable<Question>;
```

Initializing `tableDataSource` beforehand:
```typescript
tableDataSource = new MatTableDataSource<Question>([]); // defaulting to empty table
```

And `destroy$` does the magic stuff of killing the hanging subscription:
```typescript
destroy$: Subject<boolean> = new Subject<boolean>();
...
ngOnDestroy() {
  this.destroy$.next(true);
  this.destroy$.unsubscribe();
}
```

---

Another way to solve this would be to make the datasource an observable:

```typescript
this.datasource$ = questionsSvc.questions$.map(q => {
  const ds = new MatTableDataSource(q));
  ds.sort = this.sort;
  ds.filter = this.filter; // we should set this somewhere in the class
  return ds;
}
```

And then in the template we just do:

```html
<table mat-table matSort [dataSource]="datasource$ | async" class="mat-elevation-z8">
```